presto-spark-base/src/main/java/com/facebook/presto/spark/planner/IterativePlanFragmenter.java

```diff
@@ -258,7 +258,17 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<FragmentProperti
             }
 
             // don't fragment
-            return context.defaultRewrite(node, context.get());
+            if (node.getScope() == LOCAL) {
```

Comment:
I would merge this above:

if(node.getScope != REMOTE_MATERIALIZED || isFragmentReadyForExecution(node)) {
    super.visitExchange(node, context);
}